tools/clang/unittests/HLSLExec/CoopVec.h

```diff
@@ -354,6 +359,520 @@ GetMatrixSrcDataType(D3D12_LINEAR_ALGEBRA_DATATYPE MatrixInterpretation) {
+    return sizeof(int8_t);
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_SINT16:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_UINT16:
+    return sizeof(int16_t);
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_SINT32:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_UINT32:
+    if (DataInterpretation == D3D12_LINEAR_ALGEBRA_DATATYPE_SINT8_T4_PACKED ||
+        DataInterpretation == D3D12_LINEAR_ALGEBRA_DATATYPE_UINT8_T4_PACKED) {
+      return sizeof(int8_t);
+    } else {
+      return sizeof(int32_t);
+    }
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT16:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT_E4M3:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT_E5M2:
+    return sizeof(DirectX::PackedVector::HALF);
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT32:
+    return sizeof(float);
+  default:
+    throw std::invalid_argument("Unsupported data type");
+  }
+}
+
+static size_t
+GetMatrixElementSize(D3D12_LINEAR_ALGEBRA_DATATYPE DataInterpretation) {
+  switch (DataInterpretation) {
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_SINT8:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_UINT8:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_SINT16:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_UINT16:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_SINT32:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_UINT32:
+    // The CPU reference matrix is always int8 for all integer
+    // interpretations. The GPU version will be converted to the destination
+    // format by ConvertLinearAlgebraMatrix.
+    return sizeof(int8_t);
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT16:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT_E4M3:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT_E5M2:
+  case D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT32:
+    // The CPU reference matrix is always FP32 for all FP interpretations.
+    // The GPU version will be converted to the destination format by
+    // ConvertLinearAlgebraMatrix.
+    return sizeof(float);
+  default:
+    throw std::invalid_argument("Unsupported data type");
+  }
+}
+
+class TestVector {
+private:
+  size_t NumVectors = 0;
+  size_t VectorSize = 0;
+  size_t ElementSize = 0;
+  size_t Stride = 0;
+  size_t TotalBytes = 0;
+  uint8_t *Buffer = nullptr;
+
+public:
+  TestVector(size_t NumVectors, size_t VectorSize, size_t ElementSize,
+             size_t Alignment = 16)
+      : NumVectors(NumVectors), VectorSize(VectorSize),
+        ElementSize(ElementSize) {
+    if (NumVectors == 0)
+      throw std::invalid_argument("NumVectors must be greater than 0");
+    if (VectorSize == 0)
+      throw std::invalid_argument("VectorSize must be greater than 0");
+    if (ElementSize == 0)
+      throw std::invalid_argument("ElementSize must be greater than 0");
+
+    const size_t VectorBytes = VectorSize * ElementSize;
+    Stride = ((VectorBytes + Alignment - 1) / Alignment) * Alignment;
+    TotalBytes = Stride * NumVectors;
+
+    void *Ptr = nullptr;
+#ifdef _MSC_VER
+    Ptr = _aligned_malloc(TotalBytes, Alignment);
+#else
+    Ptr = std::aligned_alloc(Alignment, TotalBytes);
```

Comment:
>std::aligned_alloc [](http://example.com/codeflow?start=10&length=18)

std::alligned_alloc has undefined behavior if size is not an integral multiple of alignment. So we should do a sanity check on those inputs.

[https://en.cppreference.com/w/cpp/memory/c/aligned_alloc](https://en.cppreference.com/w/cpp/memory/c/aligned_alloc)